File: AtCoder/ARC180/B2.py
from heapq import heappop, heappush

N, K = map(int, input().split())
P = list(map(int, input().split()))

heap = []

for i, p in enumerate(P):
    for j, q in enumerate(P[i + K :], i + K):
        if q < p:
            heappush(heap, (p - q, i, j, p, q))

# print(heap)

ans = []
done = [False] * (N * N)
while heap:
    _, i, j, p, q = heappop(heap)
    if P[i] != p or P[j] != q or done[i * N + j]:
        continue
    ans.append((i + 1, j + 1))
    done[i * N + j] = True
    p, q = q, p
    P[i] = p
    P[j] = q
    for jj, qq in enumerate(P[i + K :], i + K):
        if qq < p:
            heappush(heap, (p - qq, i, jj, p, qq))
    for ii, pp in enumerate(P[: j - K + 1]):
        if q < pp:
            heappush(heap, (pp - q, ii, j, pp, q))


print(len(ans))
for i, j in ans:
    print(i, j)

File: AtCoder/ARC180/B2.rs
fn read_line() -> String {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    String::from(s.trim())
}

fn read_vec() -> Vec<usize> {
    read_line().split_whitespace().map(|c| c.parse().unwrap()).collect()
}

fn main() {
    let nk = read_vec();
    let n = nk[0];
    let k = nk[1];
    let mut pp = read_vec();

    let mut heap = std::collections::BinaryHeap::new();
    for i in 0..n {
        let p = pp[i];
        for j in i + k..n {
            let q = pp[j];
            if p > q {
                heap.push((n - (p - q), i, j));
            }
        }
    }

    let mut ans = vec![];
    let mut done = vec![false; n * n];
    while let Some((d, i, j)) = heap.pop() {
        let key = i * n + j;

        let p0 = pp[i];
        let q0 = pp[j];

        if p0 < q0 || p0 - q0 > d || done[key] {
            continue;
        }
        let (p, q) = (q0, p0);
        pp[i] = p;
        pp[j] = q;
        done[key] = true;
        ans.push((i + 1, j + 1));

        for jj in i + k..n {
            let key = i * n + jj;
            let q = pp[jj];
            if !done[key] && p > q {
                heap.push((n - (p - q), i, jj));
            }
        }

        for ii in 0..=(j - k) {
            let key = ii * n + j;
            let p = pp[ii];
            if !done[key] && p > q {
                heap.push((n - (p - q), ii, j));
            }
        }
    }

    println!("{}", ans.len());
    for &(i, j) in &ans {
        println!("{i} {j}");
    }
}
